File: parse_courses.py
import pandas as pd
import re
from pathlib import Path

# Read the XLSX file, ensuring CODE is read as string
input_file = r'c:\Users\Ogochukwu\Downloads\1.xlsx'
df = pd.read_excel(input_file, dtype={'CODE': str})

print("Original CSV columns:", df.columns.tolist())
print("\nOriginal shape:", df.shape)
print("\nFirst row:")
print(df.iloc[0])

# Parse data
new_rows = []

for idx, row in df.iterrows():
    code = str(row['CODE']).strip()  # Convert to string to preserve formatting
    title = row['COURSE TITLE']
    dates_str = str(row.get('DATE', ''))
    venues_str = str(row.get('VENUE WITH FEES', ''))
    
    # Split by newlines
    dates = [d.strip() for d in dates_str.split('\n') if d.strip()]
    venues_fees = [v.strip() for v in venues_str.split('\n') if v.strip()]
    
    print(f"\n\nProcessing {code}:")
    print(f"Dates: {dates}")
    print(f"Venues/Fees: {venues_fees}")
    
    # Match each date with venue and fee
    for i, date in enumerate(dates):
        if i < len(venues_fees):
            venue_fee = venues_fees[i]
            
            # Parse "VENUE – $FEE" or "VENUE - $FEE"
            match = re.match(r'^(.+?)\s*[-–]\s*\$?([\d,\.]+)', venue_fee)
            if match:
                venue = match.group(1).strip()
                fee_str = match.group(2).strip()
                fee = float(fee_str.replace(',', ''))
                
                new_rows.append({
                    'CODE': code,
                    'COURSE TITLE': title,
                    'DATE': date,
                    'VENUE': venue,
                    'FEE': fee,
                    'CURRENCY': 'NGN'
                })
                
                print(f"  -> Date: {date}, Venue: {venue}, Fee: {fee}")

# Create new DataFrame
new_df = pd.DataFrame(new_rows)

# Convert FEE to integer (remove decimals)
new_df['FEE'] = new_df['FEE'].astype(int)

# Save to CSV
output_file = r'c:\Users\Ogochukwu\Downloads\1_parsed.csv'
new_df.to_csv(output_file, index=False, quoting=1)  # quoting=1 is QUOTE_ALL

print(f"\n\nParsed CSV created: {output_file}")
print(f"Total expanded rows: {len(new_rows)}")
print("\nFirst 5 rows:")
print(new_df.head())

File: parse_courses.php
<?php

require_once 'vendor/autoload.php';

use PhpOffice\PhpSpreadsheet\IOFactory;

$inputFile = 'c:/Users/Ogochukwu/Downloads/1.xlsx';
$spreadsheet = IOFactory::load($inputFile);
$sheet = $spreadsheet->getActiveSheet();

$rows = [];
$dateIndex = 0;
$venueIndex = 0;
$feeIndex = 0;

// Read the header
foreach ($sheet->getRowIterator(1, 1) as $row) {
    $cellIterator = $row->getCellIterator();
    $headerIndex = 0;
    foreach ($cellIterator as $cell) {
        $value = $cell->getValue();
        if (stripos($value, 'DATE') !== false) {
            $dateIndex = $headerIndex;
        } elseif (stripos($value, 'VENUE') !== false) {
            $venueIndex = $headerIndex;
        } elseif (stripos($value, 'FEE') !== false && $feeIndex === 0) {
            $feeIndex = $headerIndex;
        }
        $headerIndex++;
    }
}

// Read data rows
$newRows = [];
foreach ($sheet->getRowIterator(2) as $row) {
    $cellIterator = $row->getCellIterator();
    $rowData = [];
    $cellIndex = 0;
    foreach ($cellIterator as $cell) {
        $rowData[$cellIndex] = $cell->getValue();
        $cellIndex++;
    }
    
    // Parse dates and venues
    $datesStr = trim($rowData[$dateIndex] ?? '');
    $venuesStr = trim($rowData[$venueIndex] ?? '');
    
    // Split dates by newline
    $dates = array_filter(array_map('trim', explode("\n", $datesStr)));
    
    // Parse venues and fees from "VENUE – $FEE" format
    $venuesAndFees = array_filter(array_map('trim', explode("\n", $venuesStr)));
    
    // Create a row for each date/venue/fee combination
    foreach ($dates as $idx => $date) {
        if (isset($venuesAndFees[$idx])) {
            $venueFee = $venuesAndFees[$idx];
            
            // Extract venue and fee
            if (preg_match('/^(.+?)\s*[-–]\s*\$?([\d,\.]+)/', $venueFee, $matches)) {
                $venue = trim($matches[1]);
                $fee = floatval(str_replace(',', '', $matches[2]));
                
                $newRow = $rowData;
                $newRow[$dateIndex] = $date;
                $newRow[$venueIndex] = $venue;
                $newRow[$feeIndex] = $fee;
                
                $newRows[] = $newRow;
            }
        }
    }
}

// Write output CSV
$outputFile = 'c:/Users/Ogochukwu/Downloads/1_parsed.csv';
$handle = fopen($outputFile, 'w');

// Write header
$header = [];
foreach ($sheet->getRowIterator(1, 1) as $row) {
    $cellIterator = $row->getCellIterator();
    foreach ($cellIterator as $cell) {
        $header[] = $cell->getValue();
    }
    break;
}
fputcsv($handle, $header);

// Write rows
foreach ($newRows as $row) {
    fputcsv($handle, $row);
}
fclose($handle);

echo "Parsed CSV created: $outputFile\n";
echo "Total rows: " . count($newRows) . "\n";

// Display sample
echo "\nFirst 5 rows:\n";
$fp = fopen($outputFile, 'r');
for ($i = 0; $i < 6; $i++) {
    $row = fgetcsv($fp);
    if ($row) {
        echo implode(' | ', $row) . "\n";
    }
}
fclose($fp);
?>
